libraries/chain/combined_database.cpp

```diff
@@ -0,0 +1,609 @@
+#include <eosio/chain/backing_store/kv_context_rocksdb.hpp>
+#include <eosio/chain/backing_store/kv_context_chainbase.hpp>
+#include <eosio/chain/combined_database.hpp>
+#include <eosio/chain/kv_chainbase_objects.hpp>
+#include <eosio/chain/backing_store/db_context.hpp>
+
+namespace eosio { namespace chain {
+   combined_session::combined_session(chainbase::database& cb_database, eosio::session::undo_stack<rocks_db_type>* undo_stack)
+       : kv_undo_stack{ undo_stack } {
+      try {
+        try {
+            cb_session = std::make_unique<chainbase::database::session>(cb_database.start_undo_session(true));
+            if (kv_undo_stack) {
+              kv_undo_stack->push();
+            }
+        }
+        FC_LOG_AND_RETHROW()
+      }
+      CATCH_AND_EXIT_DB_FAILURE()
+   }
+
+   combined_session::combined_session(combined_session&& src) noexcept
+       : cb_session(std::move(src.cb_session)), kv_undo_stack(std::move(src.kv_undo_stack)) {
+      kv_undo_stack = nullptr;
```

Comment:
Should this be `src.kv_undo_stack` ? Seems some more tests are in order.

Reply:
Changed to sec.kv_info_stack. This is only used for pending_state, which only ever calls push() which doesn't (effectively) do anything with kv_undo_stack.